base/iobuffer.jl

```diff
@@ -1,26 +1,101 @@
+#  ||    |           |                    |    |
+#  |1    ptr         size                 |    maxsize (≥ lastindex)
+#  mark (zero-indexed)                    lastindex(data)
+
+# * The underlying array is always 1-indexed
+# * The IOBuffer has full control (ownership) of the underlying array.
+# * Data in 1:mark can be deleted, shifting the whole thing to the left
+#   to make room for more data, without replacing or resizing data
+
+# Internal trait object used to access unsafe constructors.
+struct UnsafeMethod end
+const unsafe_method = UnsafeMethod()
 
-# Stateful string
 mutable struct GenericIOBuffer{T<:AbstractVector{UInt8}} <: IO
-    data::T # T should support: getindex, setindex!, length, copyto!, similar, and (optionally) resize!
-    reinit::Bool # if true, data needs to be re-allocated (after take!)
+    # T should support: getindex, setindex!, length, copyto!, similar, and (optionally) resize!
+    data::T
+
+    # The user can take control of `data` out of this struct. When that happens, instead of eagerly allocating
+    # a new array, we set `.reinit` to true, and then allocate a new one when needed.
+    reinit::Bool
     readable::Bool
     writable::Bool
-    seekable::Bool # if not seekable, implementation is free to destroy (compact) past read data
-    append::Bool # add data at end instead of at pointer
-    size::Int # end pointer (and write pointer if append == true) + offset
-    maxsize::Int # fixed array size (typically pre-allocated)
-    ptr::Int # read (and maybe write) pointer + offset
-    offset::Int # offset of ptr and size from actual start of data and actual size
-    mark::Int # reset mark location for ptr (or <0 for no mark)
 
-    function GenericIOBuffer{T}(data::T, readable::Bool, writable::Bool, seekable::Bool, append::Bool,
-                                maxsize::Integer) where T<:AbstractVector{UInt8}
-        require_one_based_indexing(data)
-        return new(data, false, readable, writable, seekable, append, length(data), maxsize, 1, 0, -1)
-    end
+    # If not seekable, implementation is free to destroy (compact) data in 1:mark-1.
+    # If it IS seekable, the user may always recover any data in 1:size by seeking,
+    # so no data can be destroyed
+    seekable::Bool
+
+    # If true, write new data to the index size+1 instead of the index ptr.
+    append::Bool
+
+    # Last index of `data` that has been written to. Data in size+1:end has not yet been used,
+    # and may contain arbitrary values.
+    # This value is always in 0 : lastindex(data)
+    size::Int
+
+    # This is the maximum length that the buffer size can grow to.
+    # This value is always in 0:typemax(Int).
+    # We always have length(data) <= maxsize
+    maxsize::Int
+
+    # Data is read/written from/to ptr, except in situations where append is true, in which case
+    # data is still read from ptr, but written to size+1.
+    # This value is always in 1 : size+1
+    ptr::Int
```

Comment:
I wonder if we could make this a `MemoryRef` If possible, doing so might save a few cycles.

Reply:
That can't be done, because a MemoryRef can't point to one past the end, which ptr needs to do to signal EOF

Reply:
Oscar seems probably right that we could replace the Memory+offset fields with a single MemoryRef field, since that may express almost the same concept